Implement waku/1

cammellos opened the pull request:
### Why make the change?
This implements waku/1 which is a breaking change as waku/0 diverged
from the specs.

Tested against v1.3 both in waku/0 and waku/1 mode.

### What has changed?


* Added v1 namespace
* Changed the testing code to test from the outside rather than stubbing
p2p specific messages, so that any version specific code is under `vx`
namespaces
* Split waku vs waku/x test code
* Added a test suite that can be used for different versions

### Things still to do (probably in a separate PR)
I kept the tests we have for the versioned aspect of waku, probably we
want to add some or change other to provide the optimal test coverage, but I would do this in a separate PR.

There's also a few places where we don't check for errors in the tests as it's a bit tricky (when we run the main loop), that probably need sorting out at some point , but I didn't want to complicate the tests too much, so open to suggestions.

### Things still to do in this PR
We talked about factory pattern, currently we don't have to look into `caps` directly as we can pass directly a version and an handler to `devp2p` and it will call the right one, so it seems like it's the best place to make that decision:

```
waku.protocols = []p2p.Protocol{{
Name: v0.Name,
Version: uint(v0.Version),
Length: v0.NumberOfMessageCodes,
Run: waku.HandlePeerV0,
NodeInfo: func() interface{} {
return map[string]interface{}{
"version": v0.VersionStr,
"maxMessageSize": waku.MaxMessageSize(),
"minimumPoW": waku.MinPow(),
}
},
},
{
Name: v1.Name,
Version: uint(v1.Version),
Length: v1.NumberOfMessageCodes,
Run: waku.HandlePeerV1,
NodeInfo: func() interface{} {
return map[string]interface{}{
"version": v1.VersionStr,
"maxMessageSize": waku.MaxMessageSize(),
"minimumPoW": waku.MinPow(),
}
},
},
}
```

Where `handlePeerVx` are just calling `HandlePeer` with the right one.

```
func (w *Waku) handlePeerV0(p2pPeer *p2p.Peer, rw p2p.MsgReadWriter) error {
return w.HandlePeer(v0.NewPeer(w, p2pPeer, rw, w.logger.Named("waku/peerv0")), rw)
}

func (w *Waku) handlePeerV1(p2pPeer *p2p.Peer, rw p2p.MsgReadWriter) error {
return w.HandlePeer(v1.NewPeer(w, p2pPeer, rw, w.logger.Named("waku/peerv1")), rw)
}
```

I looked into passing a func to HandlePeer, but it seems like it's more verbose then just having separate functions.

We talked about a factory pattern, how do you think we could/should fit this into the code @Samyoul ?

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (15)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 290e1a0b | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1956/1/) | 2020-04-29 07:44:25 | ~50 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200429-074334-290e1a.zip) |
| :heavy_check_mark: | 290e1a0b | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1956/1/) | 2020-04-29 07:49:10 | ~5 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200429-074333-290e1a.zip) |
| :heavy_check_mark: | 290e1a0b | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1956/1/) | 2020-04-29 07:51:18 | ~7 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200429-074331-290e1a.aar) |
| | | | | | | |
| :heavy_check_mark: | e4a30f25 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1956/2/) | 2020-04-29 09:22:09 | ~41 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200429-092127-e4a30f.zip) |
| :heavy_check_mark: | e4a30f25 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1956/2/) | 2020-04-29 09:24:35 | ~3 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200429-092124-e4a30f.zip) |
| :heavy_check_mark: | e4a30f25 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1956/2/) | 2020-04-29 09:28:25 | ~7 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200429-092121-e4a30f.aar) |
| | | | | | | |
| :heavy_check_mark: | 3c11c14f | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1956/3/) | 2020-04-29 18:08:05 | ~37 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200429-180727-3c11c1.zip) |
| :heavy_check_mark: | 3c11c14f | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1956/3/) | 2020-04-29 18:10:06 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200429-180726-3c11c1.zip) |
| :heavy_check_mark: | 3c11c14f | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1956/3/) | 2020-04-29 18:13:52 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200429-180724-3c11c1.aar) |
| | | | | | | |
| :heavy_check_mark: | 27b20ca7 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1956/4/) | 2020-04-30 08:11:09 | ~44 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200430-081024-27b20c.zip) |
| :heavy_check_mark: | 27b20ca7 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1956/4/) | 2020-04-30 08:13:11 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200430-081022-27b20c.zip) |
| :heavy_check_mark: | 27b20ca7 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1956/4/) | 2020-04-30 08:17:32 | ~7 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200430-081021-27b20c.aar) |
| | | | | | | |
| :heavy_check_mark: | 19115cbd | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1956/5/) | 2020-04-30 09:16:52 | ~41 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200430-091611-19115c.zip) |
| :heavy_check_mark: | 19115cbd | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1956/5/) | 2020-04-30 09:18:46 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200430-091608-19115c.zip) |
| :heavy_check_mark: | 19115cbd | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1956/5/) | 2020-04-30 09:22:15 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200430-091605-19115c.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 86569af5 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1956/6/) | 2020-04-30 13:47:55 | ~41 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200430-134713-86569a.zip) |
| :heavy_check_mark: | 86569af5 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1956/6/) | 2020-04-30 13:49:47 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200430-134711-86569a.zip) |
| :heavy_check_mark: | 86569af5 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1956/6/) | 2020-04-30 13:57:08 | ~9 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200430-134708-86569a.aar) |
| | | | | | | |
| :heavy_check_mark: | 06852d44 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1956/7/) | 2020-04-30 14:26:08 | ~43 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200430-142524-06852d.zip) |
| :heavy_check_mark: | 06852d44 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1956/7/) | 2020-04-30 14:28:54 | ~3 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200430-142521-06852d.zip) |
| :heavy_check_mark: | 06852d44 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1956/7/) | 2020-04-30 14:32:11 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200430-142519-06852d.aar) |

Samyoul commented on waku/simulation_test.go:
```diff
@@ -397,31 +394,6 @@ func sendMsg(t *testing.T, expected bool, id int) {
 	}
 }
 
-func TestPeerBasic(t *testing.T) {
-	InitSingleTest()
-
-	params, err := generateMessageParams()
-	if err != nil {
-		t.Fatalf("failed generateMessageParams with seed %d.", seed)
-	}
-
-	params.PoW = 0.001
-	msg, err := common.NewSentMessage(params)
-	if err != nil {
-		t.Fatalf("failed to create new message with seed %d: %s.", seed, err)
-	}
-	env, err := msg.Wrap(params, time.Now())
-	if err != nil {
-		t.Fatalf("failed Wrap with seed %d.", seed)
-	}
-
-	p := v0.NewPeer(nil, nil, nil, nil)
-	p.Mark(env)
-	if !p.Marked(env) {
-		t.Fatalf("failed mark with seed %d.", seed)
-	}
-}
-
```
Removing whole tests make me nervous 😄 . I see this test in https://github.com/status-im/status-go/pull/1956/files#diff-da14d6cb8f94efda56c670a48d575551R76 but I don't see an equivalent test in `waku/v0`

cammellos replied:
Yes, I only moved it to `v1`, I will add it to `v0` as well

Samyoul commented on waku/v1/message.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+// DecodeResponse1 decodes response into first version of the messages response.
+func (m MultiVersionResponse) DecodeResponse1() (resp common.MessagesResponse, err error) {
+	return resp, rlp.DecodeBytes(m.Response, &resp)
+}
```
I know the original that this was copied from doesn't have any tests, but this could be a good opportunity to introduce some / one.

jakubgs replied:
At least in a separate commit. Separate PR would be better.

Samyoul commented on waku/v1/message.go:
```diff
@@ -0,0 +1,55 @@
+	}
+}
+
+func sendBundle(rw p2p.MsgWriter, bundle []*common.Envelope) (rst gethcommon.Hash, err error) {
+	data, err := rlp.EncodeToBytes(bundle)
+	if err != nil {
+		return
+	}
+	err = rw.WriteMsg(p2p.Msg{
+		Code:    messagesCode,
+		Size:    uint32(len(data)),
+		Payload: bytes.NewBuffer(data),
+	})
+	if err != nil {
+		return
+	}
+	return crypto.Keccak256Hash(data), nil
+}
```
I know the original that this was copied from doesn't have any tests, but this could be a good opportunity to introduce some / one.

Samyoul commented on waku/v1/statusoptions.go:
```diff
@@ -0,0 +1,212 @@
+	"github.com/status-im/status-go/waku/common"
+)
+
+// statusOptionKey is a current type used in StatusOptions as a key.
+type statusOptionKey uint64
```
🎉

Samyoul commented on waku/v1/statusoptions.go:
```diff
@@ -0,0 +1,212 @@
+			continue
+		}
+
+		keys := strings.Split(rlpTag, "=")
+
+		if len(keys) != 2 || keys[0] != "key" {
+			panic("invalid value of \"rlp\" tag, expected \"key=N\" where N is uint")
+		}
+		key, err := strconv.ParseUint(keys[1], 10, 64)
+		if err != nil {
+			panic("could not parse \"rlp\" key")
+		}
```
🎉

Samyoul commented:
>We talked about a factory pattern, how do you think we could/should fit this into the code @Samyoul ?

Let me look over the code and think about the best way to implement this.

jakubgs commented:
I wish this was two commits. One copying Waku v0 stuff to v1, and another changing the v1 copy.
With as it is right now it's kinda hard to tell what changed.

cammellos commented:
@jakubgs yes, that's a good idea, I can split it off in 2 commits, the changes are minimal

cammellos committed:
Create waku/1 namespace

This commit creates a waku/1 namespace and adds the code to it.
No changes in the protocol have been made (i.e waku/1 is the same as
waku/1 in this commit).

cammellos committed:
Implement waku/1

Why make the change?

This implements waku/1 which is a breaking change as waku/0 diverged
from the specs.

What has changed?

- Added v1 namespace
- Changed the testing code to test from the outside rather than stubbing
p2p specific messages, so that any version specific code is under `vx`
namespaces
- Split waku vs waku/x test code
- Added a test suite that can be used for different versions

Things still to do

I kept the tests we have for the versioned aspect of waku, probably we
want to add some or change other to provide the optimal test coverage.

cammellos commented:
@jakubgs done, the first commit wire things up, and the second implements the specific `waku/1` changes, should be clearer now.

cammellos committed:
Add tests for sendBundle and fix flaky tests

cammellos commented:
@Samyoul @jakubgs I have added the tests (last commit, `message.go` had actually one test that was testing the two main functions, I have added one for `sendBundle`). It's ready to be reviewed again I believe.
Once is ok we can merge https://github.com/status-im/status-go/pull/1957 first and I can rebase against it.

Samyoul commented:
@cammellos Ok, I'm taking a look now. I've got some flaky test issues also ... I wonder if they are the same problem.

jakubgs reviewed (approved):
I'm no specialist in Waku or Whisper but I didn't see anything jarring.

Samyoul commented on waku/waku.go:
```diff
@@ -151,28 +152,37 @@ func New(cfg *Config, logger *zap.Logger) *Waku {
 
 	// p2p waku sub-protocol handler
-	waku.protocol = p2p.Protocol{
+	waku.protocols = []p2p.Protocol{{
 		Name:    v0.Name,
 		Version: uint(v0.Version),
 		Length:  v0.NumberOfMessageCodes,
-		Run:     waku.HandlePeer,
+		Run:     waku.handlePeerV0,
 		NodeInfo: func() interface{} {
 			return map[string]interface{}{
 				"version":        v0.VersionStr,
 				"maxMessageSize": waku.MaxMessageSize(),
 				"minimumPoW":     waku.MinPow(),
 			}
 		},
+	},
+		{
+			Name:    v1.Name,
+			Version: uint(v1.Version),
+			Length:  v1.NumberOfMessageCodes,
+			Run:     waku.handlePeerV1,
+			NodeInfo: func() interface{} {
+				return map[string]interface{}{
+					"version":        v1.VersionStr,
+					"maxMessageSize": waku.MaxMessageSize(),
+					"minimumPoW":     waku.MinPow(),
+				}
+			},
+		},
```
My suggestion of using factory pattern was based on a fundamental misunderstanding of your eventual approach. Now looking at this final implementation I can see that the version will be selected by the consuming code.

Do we want to introduce in another PR the selection functionality (the code that selects the highest mutual version between peers)?

cammellos replied:
I think that's already handled by devp2p, it will already do that behind the scene and call the right handler (specified in the `Run` field)

Samyoul replied:
Oooh, that is fancy! Cool.

Samyoul reviewed (approved):
Looks good. This package will be so cool once it is finished with it 🌟

Samyoul commented:
https://github.com/status-im/status-go/pull/1957 is merged

Samyoul commented:
@cammellos can I make a quick change suggestion?

Filename change:

`common/helper.go` > `common/helpers.go` (This is my fault and a tpyo that was implemented on my suggestion)
`v*/statusoptions.go` > `v*/status_options.go` (obviously this applies to `v0` and `v1`)

Samyoul commented on waku/v0/message_test.go:
```diff
@@ -40,3 +41,29 @@ func TestEncodeDecodeVersionedResponse(t *testing.T) {
+
+func TestSendBundle(t *testing.T) {
+	rw1, rw2 := p2p.MsgPipe()
+	defer func() { handleError(t, rw1.Close()) }()
+	defer func() { handleError(t, rw2.Close()) }()
```
These look nicer than the way I was implementing the same functionality. Inline is so fine 😎

Samyoul commented on services/wakuext/api_test.go:
```diff
@@ -303,8 +303,8 @@ func (s *WakuNodeMockSuite) SetupTest() {
 	s.Require().NoError(err)
 	node := enode.NewV4(&pkey.PublicKey, net.ParseIP("127.0.0.1"), 1, 1)
-	peer := p2p.NewPeer(node.ID(), "1", []p2p.Cap{{"shh", 6}})
 	rw1, rw2 := p2p.MsgPipe()
+	peer := v0.NewPeer(w, p2p.NewPeer(node.ID(), "1", []p2p.Cap{{"shh", 6}}), rw2, nil)
```
Should `"shh"` be `"waku"`?

cammellos commented:
Tested against nimbus, release app (1.3, waku/0) and PR running this code (waku/1)
